src/main/java/org/manifold/compiler/front/PrimitiveNodeVertex.java

```diff
@@ -0,0 +1,178 @@
+    super(g);
+    this.portTypeEdge = portTypeEdge;
+    this.portTypeEdge.setTarget(this);
+    this.portTypeEdge.setName("port type");
+    this.attributesEdge = attributesEdge;
+    this.attributesEdge.setTarget(this);
+    this.attributesEdge.setName("attributes");
+  }
+
+  @Override
+  public String toString() {
+    if (node == null) {
+      return "primitive node (not elaborated)";
+    } else {
+      return "primitive port (" + node.toString() + ")";
+    }
+  }
+
+  private void extractPortTypes(TypeValue type, 
+      Map<String, PortTypeValue> portMap) throws TypeMismatchException {
+    if (!(type instanceof TupleTypeValue)) {
+      Map<String, TypeValue> x = new HashMap<>();
+      x.put("x", TypeTypeValue.getInstance());
+      throw new TypeMismatchException(
+          new TupleTypeValue(x),
+          type);
+    }
+    TupleTypeValue tupleType = (TupleTypeValue) type;
+    for (Map.Entry<String, TypeValue> e : tupleType.getSubtypes().entrySet()) {
+      // if any type value is NIL, don't construct a port
+      if (e.getValue().equals(NilTypeValue.getInstance())) {
+        continue;
+      }
+      if (!(e.getValue() instanceof PortTypeValue)) {
+        // we would like to throw a TypeMismatchException
+        throw new UndefinedBehaviourError(
+            "attempt to construct port with non-port type value");
+      }
+      PortTypeValue portType = (PortTypeValue) e.getValue();
+      if (portMap.containsKey(e.getKey())) {
+        throw new UndefinedBehaviourError(
+            "duplicate port '" + e.getKey() + "' on node");
+      } else {
+        portMap.put(e.getKey(), portType);
+      }
+    }
+  }
+  
+  public void elaborate() throws Exception {
+    if (node != null) {
+      return;
+    }
+    Map<String, PortTypeValue> portTypeMap = new HashMap<>();
+    ExpressionVertex portTypeVertex = portTypeEdge.getSource();
+    portTypeVertex.elaborate();
+    // must be (Type) -> (Type)
+    TypeValue typeConstructor = new FunctionTypeValue(
+        TypeTypeValue.getInstance(), TypeTypeValue.getInstance());
+    if (!(portTypeVertex.getValue() instanceof FunctionTypeValue)) {
+      throw new TypeMismatchException(
+          typeConstructor, portTypeVertex.getValue());
+    }
+    FunctionTypeValue portType = (FunctionTypeValue) portTypeVertex.getValue(); 
+    if (!(portType.getInputType().isSubtypeOf(TypeTypeValue.getInstance()))) {
+      throw new TypeMismatchException(
+          typeConstructor, portType);
+    }
+    if (!(portType.getOutputType().isSubtypeOf(TypeTypeValue.getInstance()))) {
+      throw new TypeMismatchException(
+          typeConstructor, portType);
+    }
+    
+    // check that both the input type and output type are tuple typevalues
+    // whose entries are PortTypeValue, then extract the identifier-typevalue
+    // pairs and add them to the port type map
+    extractPortTypes(portType.getInputType(), portTypeMap);
```

Comment:
nit: `portTypeMap.addAll(extractPortTypes(portType.getInputType())` instead of side effect on the passed in parameter would be more explicit & clear